platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/AddressBookInitializer.java

```diff
@@ -0,0 +1,262 @@
+        try {
+            Files.createDirectories(addressBookDirectoryPath);
+        } catch (IOException e) {
+            logger.error(EXCEPTION.getMarker(), "Not able to create directory: {}", addressBookDirectoryPath, e);
+            addressBookDirectoryPath = null;
+        }
+        this.pathToAddressBookDirectory = addressBookDirectoryPath;
+
+        if (signedState != null) {
+            this.stateAddressBook = signedState.getAddressBook();
+            this.stateSwirldState = signedState.getSwirldState();
+            SoftwareVersion stateVersion;
+            try {
+                stateVersion = signedState
+                        .getState()
+                        .getPlatformState()
+                        .getPlatformData()
+                        .getCreationSoftwareVersion();
+            } catch (Exception e) {
+                logger.error(EXCEPTION.getMarker(), "Unable to retrieve software version from signed state.");
+                stateVersion = null;
+            }
+            this.stateSoftwareVersion = stateVersion;
+        } else {
+            this.stateAddressBook = null;
+            this.stateSwirldState = null;
+            this.stateSoftwareVersion = null;
+        }
+        initialAddressBook = initialize();
+    }
+
+    public AddressBook getInitialAddressBook() {
+        return initialAddressBook;
+    }
+
+    /**
+     * Determines the address book to use.  If there is no saved platform state, the config address book stake is
+     * updated by the swirld application state.  If the configured current version of the application is higher than the
+     * save state version, the swirld state is given both the config and previously saved address book to determine
+     * stake weights.  If the address book returned by the swirld application is valid, it is the initial address book
+     * to use, otherwise the configuration address book is the one to use.  All three address books, the configuration
+     * address book, the save state address book, and the new address book to use are recorded in the used address book
+     * file.
+     *
+     * @return the address book to use in the platform.
+     */
+    private AddressBook initialize() {
+        AddressBook candidateAddressBook = null;
+        if (useConfigAddressBook) {
+            // configuration is overriding to force use of configuration address book.
+            candidateAddressBook = configAddressBook;
+        } else if (stateSoftwareVersion == null || stateAddressBook == null || stateSwirldState == null) {
+            logger.info(
+                    STARTUP.getMarker(),
+                    "The candidateAddressBook is set to genesisSwirldState.updateStake(configAddressBook, null)."
+                            + " Either the stateVersion or the stateAddressBook or the stateSwirldState are null.");
+            final SwirldState genesisState = genesisSupplier.get();
+            candidateAddressBook =
```

Comment:
what's the point of calling `updateStake` if the second address book param is null?

Reply:
It signals to the SwirldState that we are in a genesis condition as there was no signed state address book available.